tests/bdd/features/volume_group/create/test_create.py

```diff
@@ -0,0 +1,267 @@
+def check_replicas_different_pool():
+    """each replica should reside on a different pool."""
+    filtered_volumes = get_volume_group_volumes()
+    already_encountered_pools = set()
+    for volume in filtered_volumes:
+        replica_topology = volume.state.replica_topology
+        for replica in replica_topology.values():
+            pool = replica.pool
+            if pool in already_encountered_pools:
+                assert False
+            already_encountered_pools.add(pool)
+
+
+@then("pools can have more than one replicas all belonging to different volumes")
+def check_pool_replicas_different_volumes():
+    """pools can have more than one replicas all belonging to different volumes."""
+    filtered_volumes = get_volume_group_volumes()
+    pool_replicas = {}
+    for volume in filtered_volumes:
+        replica_topology = volume.state.replica_topology
+        for replica in replica_topology.values():
+            pool = replica.pool
+            if pool in pool_replicas:
+                if volume.spec.uuid in pool_replicas[pool]:
+                    assert False
+                else:
+                    pool_replicas[pool].append(volume.spec.uuid)
+            else:
+                pool_replicas[pool] = [volume.spec.uuid]
+
+
+# HELPER METHODS
+
+# Creates the pools based on given nodes and counts.
+def create_node_pools(node_pool_map: Dict[str, int]):
+    for node_name, num_pools in node_pool_map.items():
+        for i in range(num_pools):
+            pool_name = f"pool_{uuid4().hex}"
+            disk_name = f"disk{i+1}"
+            ApiClient.pools_api().put_node_pool(
+                node_name,
+                pool_name,
+                CreatePoolBody([f"malloc:///{disk_name}?size_mb=200"]),
+            )
+    ApiClient.volumes_api().put_volume(
+        NON_VG_VOLUME_UUID, CreateVolumeBody(VolumePolicy(False), 2, VOLUME_SIZE, False)
+    )
+
+
+# Creates a 3 volume volume_group based on number of replica of each volume
+# and thin provisioning enabled or not. Launches 3 threads in parallel to execute the request.
+def create_volume_group(replicas, thin):
+    def put_volume(volume_uuid):
+        ApiClient.volumes_api().put_volume(
+            volume_uuid,
+            CreateVolumeBody(
+                VolumePolicy(False),
+                replicas=replicas,
+                size=VOLUME_SIZE,
+                thin=thin,
+                volume_group=VG_PARAM,
+            ),
+        )
+
+    volume_uuids = [VG_VOLUME_UUID_1, VG_VOLUME_UUID_2, VG_VOLUME_UUID_3]
+
+    with ThreadPoolExecutor(max_workers=len(volume_uuids)) as executor:
+        futures = [
+            executor.submit(put_volume, volume_uuid) for volume_uuid in volume_uuids
+        ]
+        for future in as_completed(futures):
+            future.result()
+
+
+# Get the volume group volumes based on uuids
+def get_volume_group_volumes():
+    volumes = ApiClient.volumes_api().get_volumes()
+    return [
```

Comment:
Is this formatted properly, looks a bit confusing to me

Reply:
Yeah it is formatted.